File: src/main/java/com/owncloud/android/operations/DownloadFileOperation.java
/*
 *   ownCloud Android client application
 *
 *   @author David A. Velasco
 *   @author masensio
 *   Copyright (C) 2015 ownCloud Inc.
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License version 2,
 *   as published by the Free Software Foundation.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.owncloud.android.operations;

import android.accounts.Account;
import android.content.Context;
import android.text.TextUtils;
import android.webkit.MimeTypeMap;

import com.nextcloud.client.account.User;
import com.owncloud.android.datamodel.DecryptedFolderMetadata;
import com.owncloud.android.datamodel.FileDataStorageManager;
import com.owncloud.android.datamodel.OCFile;
import com.owncloud.android.lib.common.OwnCloudClient;
import com.owncloud.android.lib.common.network.OnDatatransferProgressListener;
import com.owncloud.android.lib.common.operations.OperationCancelledException;
import com.owncloud.android.lib.common.operations.RemoteOperation;
import com.owncloud.android.lib.common.operations.RemoteOperationResult;
import com.owncloud.android.lib.common.utils.Log_OC;
import com.owncloud.android.lib.resources.files.DownloadFileRemoteOperation;
import com.owncloud.android.utils.EncryptionUtils;
import com.owncloud.android.utils.FileStorageUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Remote DownloadOperation performing the download of a file to an ownCloud server
 */
public class DownloadFileOperation extends RemoteOperation {
    private static final String TAG = DownloadFileOperation.class.getSimpleName();

    private User user;
    private OCFile file;
    private String behaviour;
    private String etag = "";
    private String activityName;
    private String packageName;

    private Context context;
    private Set<OnDatatransferProgressListener> dataTransferListeners = new HashSet<>();
    private long modificationTimestamp;
    private DownloadFileRemoteOperation downloadOperation;

    private final AtomicBoolean cancellationRequested = new AtomicBoolean(false);

    public DownloadFileOperation(User user, OCFile file, String behaviour, String activityName,
                                 String packageName, Context context) {
        if (user == null) {
            throw new IllegalArgumentException("Illegal null user in DownloadFileOperation " +
                    "creation");
        }
        if (file == null) {
            throw new IllegalArgumentException("Illegal null file in DownloadFileOperation " +
                    "creation");
        }

        this.user = user;
        this.file = file;
        this.behaviour = behaviour;
        this.activityName = activityName;
        this.packageName = packageName;
        this.context = context;
    }

    public DownloadFileOperation(User user, OCFile file, Context context) {
        this(user, file, null, null, null, context);
    }

    public String getSavePath() {
        if (file.getStoragePath() != null) {
            File parentFile = new File(file.getStoragePath()).getParentFile();
            if (parentFile != null && !parentFile.exists()) {
                parentFile.mkdirs();
            }
            File path = new File(file.getStoragePath());  // re-downloads should be done over the original file
            if (path.canWrite() || parentFile != null && parentFile.canWrite()) {
                return path.getAbsolutePath();
            }
        }
        return FileStorageUtils.getDefaultSavePathFor(user.getAccountName(), file);
    }

    public String getTmpPath() {
        return FileStorageUtils.getTemporalPath(user.getAccountName()) + file.getRemotePath();
    }

    public String getTmpFolder() {
        return FileStorageUtils.getTemporalPath(user.getAccountName());
    }

    public String getRemotePath() {
        return file.getRemotePath();
    }

    public String getMimeType() {
        String mimeType = file.getMimeType();
        if (TextUtils.isEmpty(mimeType)) {
            try {
                mimeType = MimeTypeMap.getSingleton()
                    .getMimeTypeFromExtension(
                        file.getRemotePath().substring(
                            file.getRemotePath().lastIndexOf('.') + 1));
            } catch (IndexOutOfBoundsException e) {
                Log_OC.e(TAG, "Trying to find out MIME type of a file without extension: " +
                    file.getRemotePath());
            }
        }
        if (mimeType == null) {
            mimeType = "application/octet-stream";
        }
        return mimeType;
    }

    public long getSize() {
        return file.getFileLength();
    }

    public long getModificationTimestamp() {
        return modificationTimestamp > 0 ? modificationTimestamp : file.getModificationTimestamp();
    }

    @Override
    protected RemoteOperationResult run(OwnCloudClient client) {
        /// perform the download
        synchronized(cancellationRequested) {
            if (cancellationRequested.get()) {
                return new RemoteOperationResult(new OperationCancelledException());
            }
        }

        RemoteOperationResult result;
        File newFile;
        boolean moved;

        /// download will be performed to a temporal file, then moved to the final location
        File tmpFile = new File(getTmpPath());

        String tmpFolder =  getTmpFolder();

        downloadOperation = new DownloadFileRemoteOperation(file.getRemotePath(), tmpFolder);
        Iterator<OnDatatransferProgressListener> listener = dataTransferListeners.iterator();
        while (listener.hasNext()) {
            downloadOperation.addDatatransferProgressListener(listener.next());
        }
        result = downloadOperation.execute(client);

        if (result.isSuccess()) {
            modificationTimestamp = downloadOperation.getModificationTimestamp();
            etag = downloadOperation.getEtag();
            newFile = new File(getSavePath());
            if (!newFile.getParentFile().exists() && !newFile.getParentFile().mkdirs()) {
                Log_OC.e(TAG, "Unable to create parent folder " + newFile.getParentFile().getAbsolutePath());
            }

            // decrypt file
            if (file.isEncrypted()) {
                FileDataStorageManager fileDataStorageManager = new FileDataStorageManager(user, context.getContentResolver());

                OCFile parent = fileDataStorageManager.getFileByPath(file.getParentRemotePath());

                DecryptedFolderMetadata metadata = EncryptionUtils.downloadFolderMetadata(parent,
                                                                                          client,
                                                                                          context,
                                                                                          user.toPlatformAccount());

                if (metadata == null) {
                    return new RemoteOperationResult(RemoteOperationResult.ResultCode.METADATA_NOT_FOUND);
                }
                byte[] key = EncryptionUtils.decodeStringToBase64Bytes(metadata.getFiles()
                        .get(file.getEncryptedFileName()).getEncrypted().getKey());
                byte[] iv = EncryptionUtils.decodeStringToBase64Bytes(metadata.getFiles()
                        .get(file.getEncryptedFileName()).getInitializationVector());
                byte[] authenticationTag = EncryptionUtils.decodeStringToBase64Bytes(metadata.getFiles()
                        .get(file.getEncryptedFileName()).getAuthenticationTag());

                try {
                    byte[] decryptedBytes = EncryptionUtils.decryptFile(tmpFile, key, iv, authenticationTag);

                    try (FileOutputStream fileOutputStream = new FileOutputStream(tmpFile)) {
                        fileOutputStream.write(decryptedBytes);
                    }
                } catch (Exception e) {
                    return new RemoteOperationResult(e);
                }
            }
            moved = tmpFile.renameTo(newFile);
            newFile.setLastModified(file.getModificationTimestamp());
            if (!moved) {
                result = new RemoteOperationResult(RemoteOperationResult.ResultCode.LOCAL_STORAGE_NOT_MOVED);
            }
        }
        Log_OC.i(TAG, "Download of " + file.getRemotePath() + " to " + getSavePath() + ": " +
                result.getLogMessage());

        return result;
    }

    public void cancel() {
        cancellationRequested.set(true);   // atomic set; there is no need of synchronizing it
        if (downloadOperation != null) {
            downloadOperation.cancel();
        }
    }


    public void addDatatransferProgressListener (OnDatatransferProgressListener listener) {
        synchronized (dataTransferListeners) {
            dataTransferListeners.add(listener);
        }
    }

    public void removeDatatransferProgressListener(OnDatatransferProgressListener listener) {
        synchronized (dataTransferListeners) {
            dataTransferListeners.remove(listener);
        }
    }

    @Deprecated
    public Account getAccount() {
        return this.user.toPlatformAccount();
    }

    public User getUser() {
        return this.user;
    }

    public OCFile getFile() {
        return this.file;
    }

    public String getBehaviour() {
        return this.behaviour;
    }

    public String getEtag() {
        return this.etag;
    }

    public String getActivityName() {
        return this.activityName;
    }

    public String getPackageName() {
        return this.packageName;
    }
}

File: app/src/main/java/com/owncloud/android/operations/DownloadFileOperation.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2026 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */
package com.owncloud.android.operations

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.webkit.MimeTypeMap
import com.nextcloud.client.account.User
import com.nextcloud.utils.extensions.showToast
import com.nextcloud.utils.extensions.toNextcloudClient
import com.owncloud.android.R
import com.owncloud.android.datamodel.ArbitraryDataProviderImpl
import com.owncloud.android.datamodel.FileDataStorageManager
import com.owncloud.android.datamodel.OCFile
import com.owncloud.android.datamodel.e2e.v1.decrypted.DecryptedFolderMetadataFileV1
import com.owncloud.android.datamodel.e2e.v2.decrypted.DecryptedFolderMetadataFile
import com.owncloud.android.lib.common.OwnCloudClient
import com.owncloud.android.lib.common.network.OnDatatransferProgressListener
import com.owncloud.android.lib.common.operations.OperationCancelledException
import com.owncloud.android.lib.common.operations.RemoteOperation
import com.owncloud.android.lib.common.operations.RemoteOperationResult
import com.owncloud.android.lib.common.utils.Log_OC
import com.owncloud.android.lib.resources.files.DownloadFileRemoteOperation
import com.owncloud.android.utils.EncryptionUtils
import com.owncloud.android.utils.FileExportUtils
import com.owncloud.android.utils.FileStorageUtils
import java.io.File
import java.io.IOException
import java.lang.ref.WeakReference
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import javax.crypto.Cipher

@Suppress("LongParameterList", "ReturnCount")
class DownloadFileOperation(
    val user: User,
    val file: OCFile,
    val behaviour: String?,
    val activityName: String?,
    val packageName: String?,
    context: Context?,
    var downloadType: DownloadType?
) : RemoteOperation<Unit>() {

    var etag: String? = ""
        private set

    private val context = WeakReference(context)
    private val dataTransferListeners = ConcurrentHashMap.newKeySet<OnDatatransferProgressListener>()
    private var timestampForModification: Long = 0
    private val cancellationRequested = AtomicBoolean(false)
    private val mainThreadHandler = Handler(Looper.getMainLooper())

    constructor(user: User, file: OCFile, context: Context?) : this(
        user,
        file,
        null,
        null,
        null,
        context,
        DownloadType.DOWNLOAD
    )

    fun isMatching(accountName: String?, fileId: Long) = file.fileId == fileId && user.accountName == accountName

    fun cancelMatchingOperation(accountName: String?, fileId: Long) {
        if (isMatching(accountName, fileId)) cancel()
    }

    val savePath: String
        get() {
            file.storagePath?.let { storagePath ->
                val parentFile = File(storagePath).parentFile
                parentFile?.takeIf { !it.exists() }?.runCatching {
                    Files.createDirectories(toPath())
                }?.onFailure {
                    return FileStorageUtils.getDefaultSavePathFor(user.accountName, file)
                }
                val path = File(storagePath)
                if (path.canWrite() || parentFile?.canWrite() == true) return path.absolutePath
            }
            return FileStorageUtils.getDefaultSavePathFor(user.accountName, file)
        }

    val tmpPath: String get() = FileStorageUtils.getTemporalPath(user.accountName) + file.remotePath
    val tmpFolder: String get() = FileStorageUtils.getTemporalPath(user.accountName)
    val remotePath: String get() = file.remotePath

    val mimeType: String
        get() {
            var mimeType = file.mimeType
            if (mimeType.isNullOrEmpty()) {
                runCatching {
                    mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(
                        file.remotePath.substring(file.remotePath.lastIndexOf('.') + 1)
                    )
                }.onFailure {
                    Log_OC.e(TAG, "Trying to find out MIME type of a file without extension: ${file.remotePath}")
                }
            }
            return mimeType ?: "application/octet-stream"
        }

    val size: Long get() = file.fileLength

    fun getModificationTimestamp() =
        if (timestampForModification > 0) timestampForModification else file.modificationTimestamp

    @Suppress("DEPRECATION")
    @Deprecated("Deprecated in Java")
    override fun run(client: OwnCloudClient): RemoteOperationResult<Unit> {
        synchronized(cancellationRequested) {
            if (cancellationRequested.get()) return RemoteOperationResult(OperationCancelledException())
        }

        if (!FileStorageUtils.isValidExtFilename(file.fileName)) {
            mainThreadHandler.post { context.get()?.showToast(R.string.download_download_invalid_local_file_name) }
            return RemoteOperationResult(RemoteOperationResult.ResultCode.INVALID_CHARACTER_IN_NAME)
        }

        val operationContext = context.get()
            ?: return RemoteOperationResult(RemoteOperationResult.ResultCode.UNKNOWN_ERROR)

        val tmpFile = File(tmpPath)
        val (downloadOp, downloadResult) = executeDownload(client, operationContext)

        if (!downloadResult.isSuccess) return downloadResult

        timestampForModification = downloadOp.modificationTimestamp
        etag = downloadOp.etag

        if (file.isEncrypted) {
            handleDecryption(client, operationContext, tmpFile)?.let { return it }
        }

        val result = when (downloadType) {
            DownloadType.DOWNLOAD if !file.isEncrypted -> handleFileMove(tmpFile, downloadResult)
            DownloadType.EXPORT -> handleExport(operationContext, tmpFile, downloadResult)
            else -> downloadResult
        }

        Log_OC.i(TAG, "Download of ${file.remotePath} to $savePath: ${result.logMessage}")
        return result
    }

    private data class DownloadResult(
        val operation: DownloadFileRemoteOperation,
        val result: RemoteOperationResult<Unit>
    )

    @Suppress("UNCHECKED_CAST")
    private fun executeDownload(client: OwnCloudClient, operationContext: Context): DownloadResult {
        val operation = DownloadFileRemoteOperation(file.remotePath, tmpFolder, file.fileLength).also { op ->
            if (downloadType == DownloadType.DOWNLOAD) {
                dataTransferListeners.forEach { op.addProgressListener(it) }
            }
        }
        val result = operation.execute(client.toNextcloudClient(operationContext)) as RemoteOperationResult<Unit>
        return DownloadResult(operation, result)
    }

    private data class EncryptionKeys(val key: String?, val nonce: String?, val authTag: String?)

    private fun extractEncryptionKeys(metadata: Any): EncryptionKeys? = when (metadata) {
        is DecryptedFolderMetadataFile -> metadata.metadata.files[file.encryptedFileName]?.let {
            EncryptionKeys(it.key, it.nonce, it.authenticationTag)
        }

        is DecryptedFolderMetadataFileV1 -> metadata.files[file.encryptedFileName]?.let {
            EncryptionKeys(it.encrypted.key, it.initializationVector, it.authenticationTag)
        }

        else -> null
    }

    @Suppress("DEPRECATION")
    private fun handleDecryption(
        client: OwnCloudClient,
        operationContext: Context,
        tmpFile: File
    ): RemoteOperationResult<Unit>? {
        val fileDataStorageManager = FileDataStorageManager(user, operationContext.contentResolver)
        val parent = fileDataStorageManager.getFileByEncryptedRemotePath(file.parentRemotePath)
        val metadata = EncryptionUtils.downloadFolderMetadata(parent, client, operationContext, user)
            ?: return RemoteOperationResult(RemoteOperationResult.ResultCode.METADATA_NOT_FOUND)

        val (keyString, nonceString, authenticationTagString) = extractEncryptionKeys(metadata)
            ?: return RemoteOperationResult(RemoteOperationResult.ResultCode.METADATA_NOT_FOUND)

        val key = EncryptionUtils.decodeStringToBase64Bytes(keyString)
        val iv = EncryptionUtils.decodeStringToBase64Bytes(nonceString)

        return runCatching {
            val cipher = EncryptionUtils.getCipher(Cipher.DECRYPT_MODE, key, iv)
            EncryptionUtils.decryptFile(
                cipher,
                tmpFile,
                File(savePath),
                authenticationTagString,
                ArbitraryDataProviderImpl(operationContext),
                user
            )
        }.fold(
            onSuccess = { null },
            onFailure = { RemoteOperationResult(it as? Exception ?: Exception(it)) }
        )
    }

    private fun handleFileMove(tmpFile: File, currentResult: RemoteOperationResult<Unit>): RemoteOperationResult<Unit> {
        val newFile = File(savePath)
        newFile.parentFile?.takeIf { !it.exists() }?.also {
            if (!it.mkdirs()) Log_OC.e(TAG, "Unable to create parent folder ${it.absolutePath}")
        }

        val tempFilePath = tmpFile.toPath()
        val newFilePath = newFile.toPath()

        Log_OC.d(TAG, "trying to move: $tempFilePath to $newFilePath")

        return try {
            Files.move(tempFilePath, newFilePath, StandardCopyOption.REPLACE_EXISTING)
            newFile.setLastModified(file.modificationTimestamp)
                .also { Log_OC.d(TAG, "Last modified set: $it") }
            currentResult
        } catch (e: IOException) {
            Log_OC.e(TAG, "Failed to move file to ${newFile.absolutePath}: ${e.message}")
            RemoteOperationResult(RemoteOperationResult.ResultCode.LOCAL_STORAGE_NOT_MOVED)
        }
    }

    private fun handleExport(
        operationContext: Context,
        tmpFile: File,
        currentResult: RemoteOperationResult<Unit>
    ): RemoteOperationResult<Unit> {
        FileExportUtils().exportFile(file.fileName, file.mimeType, operationContext.contentResolver, null, tmpFile)
        if (!tmpFile.delete()) Log_OC.e(TAG, "Deletion of ${tmpFile.absolutePath} failed!")
        return currentResult
    }

    fun cancel() {
        cancellationRequested.set(true)
    }

    fun addProgressListener(listener: OnDatatransferProgressListener) {
        dataTransferListeners.add(listener)
    }

    fun removeProgressListener(listener: OnDatatransferProgressListener) {
        dataTransferListeners.remove(listener)
    }

    companion object {
        private val TAG = DownloadFileOperation::class.java.simpleName
    }
}
